File: solutions/swapNodesInPairs-24/swapNodesInPairs-24.cpp

/**********************************************************************************
*
* Given a linked list, swap every two adjacent nodes and return its head.
*
* Input: head = [1,2,3,4]
* Output: [2,1,4,3]
*
*
**********************************************************************************/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {

  /*
  *   [Normal LinkedList Operation]
  *
  *   Time complexity : O(n).
  *
  *   Space Complexity: O(1).
  */

public:
    ListNode* swapPairs(ListNode* head) {
        
        ListNode dummy;
        dummy.next = head;
        
        ListNode* curr = &dummy;
        while(curr->next && curr->next->next) {
            
            ListNode* first = curr->next;
            ListNode* second = curr->next->next;
            
            first->next = second->next;
            second->next = first;
            curr->next = second;
            curr = first;
        }
        
        return dummy.next;
    }
};

File: solutions/swapNodesInPairs-24/swapNodesInPairs-24.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        
        dummy = ListNode()
        dummy.next = head
        
        curr = dummy
        while curr.next and curr.next.next:
            first = curr.next
            second = curr.next.next
            
            first.next = second.next
            second.next = first
            curr.next = second
            curr = first
            
        return dummy.next
        
        